File: src/main/com/intellij/lang/jsgraphql/schema/builder/GraphQLScalarTypeCompositeDefinition.java
package com.intellij.lang.jsgraphql.schema.builder;

import com.intellij.lang.jsgraphql.types.language.Directive;
import com.intellij.lang.jsgraphql.types.language.ScalarTypeDefinition;
import com.intellij.lang.jsgraphql.types.language.ScalarTypeExtensionDefinition;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class GraphQLScalarTypeCompositeDefinition
  extends GraphQLExtendableCompositeDefinition<ScalarTypeDefinition, ScalarTypeExtensionDefinition> {

  @NotNull
  @Override
  protected ScalarTypeDefinition mergeDefinitions(@NotNull List<ScalarTypeDefinition> sourceDefinitions) {
    List<Directive> directives = new ArrayList<>();

    for (ScalarTypeDefinition definition : sourceDefinitions) {
      directives.addAll(definition.getDirectives());
    }

    ScalarTypeDefinition definition = ContainerUtil.getFirstItem(sourceDefinitions);
    return definition.transform(builder ->
                                  builder
                                    .directives(directives)
                                    .sourceNodes(sourceDefinitions)
    );
  }
}

File: src/main/com/intellij/lang/jsgraphql/schema/builder/GraphQLScalarTypeCompositeDefinition.kt
package com.intellij.lang.jsgraphql.schema.builder

import com.intellij.lang.jsgraphql.types.language.Directive
import com.intellij.lang.jsgraphql.types.language.ScalarTypeDefinition
import com.intellij.lang.jsgraphql.types.language.ScalarTypeExtensionDefinition

class GraphQLScalarTypeCompositeDefinition
  : GraphQLExtendableCompositeDefinition<ScalarTypeDefinition, ScalarTypeExtensionDefinition>() {

  override fun mergeDefinitions(sourceDefinitions: List<ScalarTypeDefinition>): ScalarTypeDefinition {
    val directives = mutableListOf<Directive>()

    for (definition in sourceDefinitions) {
      directives.addAll(definition.directives)
    }

    return sourceDefinitions.first().transform { builder ->
      builder
        .directives(directives)
        .sourceNodes(sourceDefinitions)
    }
  }
}
